File: Bedlam.ts

import _ from 'lodash'
import { createSVGWindow } from 'svgdom'
import { Box, SVG, Svg, registerWindow } from '@svgdotjs/svg.js'
import xmlFormat from 'xml-formatter'

const log = console.log

// Cube is SIDE x SIDE x SIDE
const SIDE = 4

type Encoded = bigint

// All bits of cube set
const FULL: Encoded = 0xffffffffffffffffn

// Empty cube
const EMPTY: Encoded = 0n

type Block = [number, number, number]
type Shape = Block[]

// Array of shape num x blocks x {x,y,z}
const shapes: Shape[] = [
	[[0,0,0],[0,1,0],[0,2,0],[1,1,0],[1,1,1]],
	[[0,0,0],[0,1,0],[0,2,0],[1,0,0],[0,1,1]],
	[[0,0,0],[0,1,0],[0,2,0],[1,0,1],[0,0,1]],
	[[0,0,0],[0,1,0],[0,2,0],[0,0,1],[1,0,0]],
	[[0,0,0],[0,1,0],[0,1,1],[1,1,1],[1,2,1]],
	[[0,0,0],[0,1,0],[0,1,1],[0,2,1],[1,1,1]],
	[[0,1,0],[1,0,0],[1,1,0],[1,2,0],[2,1,0]],
	[[0,0,0],[0,1,0],[0,1,1],[1,1,1]],
	[[0,0,0],[0,1,0],[0,2,0],[1,1,0],[0,1,1]],
	[[0,1,0],[0,2,0],[1,0,0],[1,1,0],[2,1,0]],
	[[0,0,0],[0,1,0],[1,0,0],[0,1,1],[0,2,1]],
	[[0,0,0],[0,1,0],[0,2,0],[1,0,0],[0,2,1]],
	[[0,1,0],[0,2,0],[1,0,0],[1,1,0],[2,0,0]],
]

enum Axis {X=0, Y=1, Z=2}

// Rotate block around axis
function rotateBlock ([x, y, z]: Block , axis: Axis): Block {
	switch (axis) {
		case Axis.X: return [x, z, -y] 
		case Axis.Y: return [-z, y, x]
		case Axis.Z: return [y, -x, z]
	}
}

// Rotate shape about axis
function rotateShape (shape: Shape, axis: Axis): Shape {
	return shape.map(block => rotateBlock(block, axis))
}

// Add all rotations about X axis
function xAxisRotations(shape: Shape): Shape[] {
	const r1 = rotateShape(shape, Axis.X)
	const r2 = rotateShape(r1, Axis.X)
	const r3 = rotateShape(r2, Axis.X)
	return [shape, r1, r2, r3]
}

// Return all 6x4 possible rotations of the shape
function translate (shape: Shape): Shape[]  {

	// Get all y axis translations
	const t1 = rotateShape(shape, Axis.Y)
	const t2 = rotateShape(t1, Axis.Y)
	const t3 = rotateShape(t2, Axis.Y)

	// Add missing z axis translations
	const t4 = rotateShape(shape, Axis.Z)
	const t5 = rotateShape(t2, Axis.Z)
	
	return [shape, t1, t2, t3, t4, t5].flatMap(xAxisRotations)
}

function axisRange(shape: Shape, axis: Axis): [number, number] {
	const aMin = _.min(shape.map(b => b[axis]))!
	const aMax = _.max(shape.map(b => b[axis]))!
	return [aMin, 4-(aMax-aMin)]
}

function encodeBlock([x, y, z]: Block): Encoded {
	return (1n << BigInt(x*SIDE*SIDE+y*SIDE+z))
}

// Encode shape array as long
function encode (shape: Shape): Encoded {
	return _.reduce(shape, (a, b) => {
		return a | encodeBlock(b)
	}, 0n)
}

// Return shape shifted by x,y,z amounts
function shift(shape: Shape, dx: number, dy: number, dz: number): Shape {
	const os: Shape =  shape.map(([x, y, z]) => [x+dx, y+dy, z+dz])
	return os
}

// Return all positions the shape could occupy in the cube
function allShapeEncodings (shape: Shape): Encoded[] {
	
	const [xMin, xWid] = axisRange(shape,Axis.X)
	const [yMin, yWid] = axisRange(shape,Axis.Y)
	const [zMin, zWid] = axisRange(shape,Axis.Z)

	// Add all encoded positions into array
	return _.range(0, xWid).flatMap (dx => 
		_.range(0, yWid).flatMap (dy => 
			_.range(0, zWid).map (dz => 
				encode(shift(shape, dx-xMin, dy-yMin, dz-zMin))
			)
		)
	)
}

function searchPiece(cube: Encoded, positions: Encoded[], remaining: Encoded[][], placed: Encoded[]): Encoded[] {
	if (_.isEmpty(positions)) return []
	const position = _.head(positions)!
	const result = search(cube | position, remaining, [...placed, position])
	if (!_.isEmpty(result)) return result
	return searchPiece(cube, _.tail(positions)!, remaining, placed)
}

// Work through each shape discarding those that clash with a filled position
function search(cube: Encoded, remaining: Encoded[][], placed: Encoded[]): Encoded[] {
	if (cube === FULL) {return placed}
	const others = _.tail(remaining)!
	const candidates = _.head(remaining)!.filter((c) => {
		if ((c & cube) !== EMPTY) {return false}
		return !prune(c | cube, others)
	})
	return searchPiece(cube, candidates, others, placed)	
}

const e2s = (encoded: Encoded) => encoded.toString(2).padStart(64, "0")

function logEnc(encoded: Encoded, name: string = "") {
	log(name.padEnd(8), e2s(encoded))
}

// Prune a search if either
// - There is a piece that can no longer fit
// - There is an enclosed hole in the cube
function prune(cube: Encoded, remaining: Encoded[][]): boolean {	
	var testCube = cube
	var allFit = remaining.findIndex((r) => {
		const candidates = r.filter((s) => (cube & s) === EMPTY)
		candidates.forEach(c => testCube |= c)
		return _.isEmpty(candidates)
	}) === -1
	return (!allFit) || (testCube !== FULL)
}

function allTranslatedShapeEncodings(s: Shape): Encoded[] {
	const set = new Set(translate(s).flatMap(t => allShapeEncodings(t)))
	return [...set]
}

const allShapeTranslations: Encoded[][] = shapes.map(allTranslatedShapeEncodings)


// Decode encoded back to shape
function decode(shape: Encoded): Shape {
	return _.range(0, 4).flatMap (x => 
		_.range(0, 4).flatMap (y => 
			_.range(0, 4).flatMap (z => {
				const b: Block = [x, y, z]
				const e = encodeBlock(b)
				return ((e & shape) === e) ? [b] : []
			})
		)
	)
}
	
// Render blocks as SVG
function toSVG(solution: Shape[]): string {

	const svg: Svg = SVG()

	const box: Box = new Box('0 0 5000 4000')

	svg.width(1000).height(800).viewbox(box)
	const panel1 = svg.defs().polygon([[0, 0], [0, 100], [86.6, 50], [86.6, -50]])
	const box1 = svg.defs().group()
		.add(svg.use(panel1).rotate(-120, 0, 0))
		.add(svg.use(panel1).rotate(0, 0, 0))
		.add(svg.use(panel1).rotate(120, 0, 0))
	const block1 = svg.defs().use(box1).fill("pink")
	const cube1 = svg.defs().use(box1).stroke({dasharray: "1,3"}).rotate(180, 0, 0).translate(0, -100).scale(4).translate(0,-200)
	
	// Write shape definitions
	const shapes1 = solution.map((shape, i) => {
		const sorted = _.sortBy(shape, ([x,y,z]) => -(x+y-100*z))
		const shape1 = svg.defs().group()
		sorted.forEach(([x, y, z]) => {
			const tx = 86.60254 * (x-y)
			const ty = -50 * (x+y) - 100 * z
			shape1.add(svg.use(block1).translate(tx, ty))
		})
		return shape1
	})

	const main1 = svg.group().fill("none").stroke("black")

	shapes1.map((s, i) => {
		const x = 800 + (i % 5) * 800
		const y = 1000 + Math.floor(i / 5) * 1000
		main1.use(cube1).move(x, y)
		main1.text(i.toString()).font({'font-size': 60}).move(x, y+150)
		main1.use(s).move(x, y)
	})
	
	svg.svg()
	return xmlFormat(svg.svg())

}

function registerSvgWindow() {
	const window = createSVGWindow()
	const document = window.document
	registerWindow(window, document)	
}

export const exampleSolution = [
	3146001n, 131682n, 9223528167505920000n, 39969481052651520n, 71339008n, 1152975380693385216n,
	549898944640n, 838864896n, 281556581154816n, 70371965427712n, 891712863658311680n, 292058038284n,
	7138205409382236160n
]

async function render(solution: Encoded[], file: string) {	
	registerSvgWindow()
	const shapes = solution.map(decode)
	const svg = toSVG(shapes)
	await Bun.write(file, svg)	
	log(`SVG written to ${file}`)
}

function run() {
	const start = Date.now()
	log("Running (takes around 20s) ...")
	const solution = search(EMPTY, allShapeTranslations, [])	// Takes around 20s
	const end = Date.now()
	log("=== Result ===")
	solution.forEach((p) => logEnc(p))
	log("Time taken", Math.round((end-start)/1000), "seconds")
	render(solution, "solution.svg")
}

run()

File: java/bedlam/Bedlam.java
package bedlam;

import java.io.FileWriter;
import java.io.Writer;
import java.util.Arrays;
import java.util.HashSet;
import java.util.Set;

public class Bedlam {

    // In two's complements all bits set is -1 (full cube)
    final static long FULL = -1l;

    // Array of shape num x blocks x {x,y,z}
    static final int [][][] shapes = {
        {{0,0,0},{0,1,0},{0,2,0},{1,1,0},{1,1,1}},
        {{0,0,0},{0,1,0},{0,2,0},{1,0,0},{0,1,1}},
        {{0,0,0},{0,1,0},{0,2,0},{1,0,1},{0,0,1}},
        {{0,0,0},{0,1,0},{0,2,0},{0,0,1},{1,0,0}},
        {{0,0,0},{0,1,0},{0,1,1},{1,1,1},{1,2,1}},
        {{0,0,0},{0,1,0},{0,1,1},{0,2,1},{1,1,1}},
        {{0,1,0},{1,0,0},{1,1,0},{1,2,0},{2,1,0}},
        {{0,0,0},{0,1,0},{0,1,1},{1,1,1}},
        {{0,0,0},{0,1,0},{0,2,0},{1,1,0},{0,1,1}},
        {{0,1,0},{0,2,0},{1,0,0},{1,1,0},{2,1,0}},
        {{0,0,0},{0,1,0},{1,0,0},{0,1,1},{0,2,1}},
        {{0,0,0},{0,1,0},{0,2,0},{1,0,0},{0,2,1}},
        {{0,1,0},{0,2,0},{1,0,0},{1,1,0},{2,0,0}}
    };

    // One possible solution (need to verify)
    static final int [][][] SOLN = {
        {{1, 2, 0},{2, 1, 1},{2, 2, 0},{2, 2, 1},{3, 2, 0}},
        {{0, 1, 0},{0, 2, 0},{0, 2, 1},{0, 3, 0},{1, 1, 0}},
        {{0, 2, 2},{1, 2, 2},{1, 3, 0},{1, 3, 1},{1, 3, 2}},
        {{0, 2, 3},{0, 3, 1},{0, 3, 2},{0, 3, 3},{1, 3, 3}},
        {{1, 1, 2},{2, 0, 3},{2, 1, 2},{2, 1, 3},{3, 0, 3}},
        {{0, 0, 3},{0, 1, 2},{0, 1, 3},{1, 1, 3},{1, 2, 3}},
        {{0, 0, 2},{1, 0, 1},{1, 0, 2},{1, 0, 3},{2, 0, 2}},
        {{2, 2, 3},{2, 3, 3},{3, 3, 2},{3, 3, 3}},
        {{1, 0, 0},{2, 0, 0},{2, 0, 1},{2, 1, 0},{3, 0, 0}},
        {{3, 1, 0},{3, 1, 1},{3, 2, 1},{3, 2, 2},{3, 3, 1}},
        {{0, 0, 0},{0, 0, 1},{0, 1, 1},{1, 1, 1},{1, 2, 1}},
        {{2, 2, 2},{2, 3, 0},{2, 3, 1},{2, 3, 2},{3, 3, 0}},
        {{3, 0, 1},{3, 0, 2},{3, 1, 2},{3, 1, 3},{3, 2, 3}}
    };

    // Generate positions
    long [][] positions;
    Bedlam() {
        int i=0;
        positions = new long[shapes.length][];
        for (int[][] shape : shapes) {
            Set<Long> posSet = new HashSet<Long>();
            for (int[][] tshape : translate(shape)) {
                addPositions(posSet, tshape);
            }
            positions[i++] = toArray(posSet);
        }
        assert i==positions.length;
    }

    // Search for solutions, return an array of offsets inside the positions array
    long solnCnt;
    final int[] solnPos = new int[shapes.length];
    void search() {
        solnCnt = System.currentTimeMillis();
        search(0l,0);
    }

    // Work through each shape discarding those that clash with a filled position
    void search(long cube, int pos) {
        if (cube == FULL) {
            System.out.println("Found solution after " + ((System.currentTimeMillis() - solnCnt)/1000) + "s");
            for (int p=0;p<solnPos.length;p++) System.out.println(toString(decode(positions[p][solnPos[p]])));
            return; // System.exit(1);
        }
        long posn[] = positions[pos];
        for (int i=0;i<posn.length;i++) {
            if ((cube & posn[i]) == 0l) {
                long newCube = cube | posn[i];
                if (!prune(newCube, pos+1)) {
                    solnPos[pos] = i;
                    search (newCube, pos+1);
                }
            }
        }
    }

    // Prune a search if either
    // - There is a piece that can no longer fit
    // - There is an enclosed hole in the cube
    boolean prune(long cube, int pos) {
        long testCube = cube;
        for (int p=pos;p<positions.length;p++) {
            boolean used = false;
            for (long shapePos : positions[p]) {
                if ((shapePos & cube) == 0l) {
                    used = true;
                    if ((testCube |= shapePos) == FULL) break;
                }
            }
            if (!used) return true;     // Piece does not fit
        }
        return !(testCube == FULL);     // Will be full if there were no holes
    }

    // Pre-allocate space or we get killed
    final int[][] offsets = new int[shapes.length][shapes.length];

    void search2() {
        solnCnt = System.currentTimeMillis();
        search2(0l,0);
    }

    // Work through each shape discarding those that clash with a filled position
    void search2(long cube, int pos) {

        long posn[] = positions[pos];
        int[] offset = offsets[pos];

        for (int i=offset[pos];i<posn.length;i++) {
            if ((cube & posn[i]) == 0l) {
                long newCube = cube | posn[i];
                if (newCube == FULL) {
                    System.out.println("Found solution after " + ((System.currentTimeMillis() - solnCnt)/1000) + "s");
                    for (int p=0;p<solnPos.length;p++) System.out.println(toString(decode(positions[p][solnPos[p]])));
                    System.exit(1);
                } else {
                    // Calculate new offsets based on newCube
                    int[] newOffset = offsets[pos+1]; // new int[offset.length];
                    boolean prune = false;
                    for (int offPos = pos+1; offPos < positions.length; offPos++) {
                        long offPosn[] = positions[offPos];
                        int o=offset[offPos];
                        for (int j=o;j<offPosn.length;j++) {
                            if ((offPosn[j] & newCube) != 0l) {
                                // Piece no longer fits so swap under offset
                                if (j>o) {
                                    long tmp = offPosn[j];
                                    offPosn[j] = offPosn[o];
                                    offPosn[o] = tmp;
                                }
                                o++;
                            }
                        }
                        if (o == offPosn.length) {
                            prune = true;
                            break;
                        }
                        newOffset[offPos] = o;
                    }

                    // Prune
                    if (!prune && !prune2(newCube, pos+1, newOffset)) {
                        solnPos[pos] = i;
                        search2 (newCube, pos+1); // , newOffset);
                    }
                }
            }
        }
    }

    // Prune a search if either
    // - There is a piece that can no longer fit
    // - There is an enclosed hole in the cube
    boolean prune2(long cube, int pos, int[] offset) {
        long testCube = cube;
        for (int p=pos;p<positions.length;p++) {
            long posn[] = positions[p];
            for (int i=offset[p];i<posn.length;i++) {
                if ((testCube |= posn[i]) == FULL) return false;
            }
        }
        return true;
    }

    // Convert long set to array
    long[] toArray(Set<Long> set) {
        int c=0;
        long[] a = new long[set.size()];
        for (long l : set) a[c++] = l;
        return a;
    }

    // Encode shape array as long
    static long encode (int [][] shape) {
        long l = 0l;
        for (int[] block : shape) {
            l |= (1l << (block[0] * (SIDE*SIDE) + block[1] * SIDE + block[2]));
        }
        return l;
    }

    // Decode shape long as array
    static int[][] decode(long shape) {
        int [][] xblocks = new int[64][];
        int c=0;
        for (int i=0;i<64;i++) {
            if ((shape & (1l << i)) != 0l) {
                long b = i;
                int x = (int) (b / (SIDE*SIDE));    b -= x * (SIDE*SIDE);
                int y = (int) (b / SIDE);           b -= (y * SIDE);
                int z = (int) (b);
                xblocks[c++] = new int[] { x, y, z };
            }
        }
        int [][] blocks = new int[c][];
        System.arraycopy(xblocks, 0, blocks, 0, c);
        return blocks;
    }

    // Render a string
    static String toString(int [][] shape) {
        StringBuilder sb = new StringBuilder();
        String c = "";
        for (int[] block : shape) {
            sb.append(c + Arrays.toString(block)); c=",";
        }
        return sb.toString();
    }

    // Helper to get max dimension
    int max(int shape[][], int dim) {
        int max = Integer.MIN_VALUE;
        for (int [] block : shape) {
            if (block[dim] > max) max = block[dim];
        }
        return max;
    }

    // Helper to get min dimension
    int min(int shape[][], int dim) {
        int min = Integer.MAX_VALUE;
        for (int [] block : shape) {
            if (block[dim] < min) min = block[dim];
        }
        return min;
    }

    // Cube is SIDE x SIDE x SIDE
    final static int SIDE = 4;

    // Return all 6x4 possible rotations of the shape
    int[][][] translate (int [][] shape) {
        int i=0;
        int [][][] trans = new int[24][][];

        // Get all y axis tranlations
        shape = rotate(shape,1); i = addTrans(shape, trans, i);
        shape = rotate(shape,1); i = addTrans(shape, trans, i);
        shape = rotate(shape,1); i = addTrans(shape, trans, i);
        shape = rotate(shape,1); i = addTrans(shape, trans, i);

        // Add missing z axis translations
        shape = rotate(shape,2); i = addTrans(shape, trans, i);
        shape = rotate(shape,2);
        shape = rotate(shape,2); i = addTrans(shape, trans, i);
        shape = rotate(shape,2);

        // Check total and return
        assert (i==trans.length);
        return trans;
    }

    // Add all rotations about an axis
    int addTrans(int [][] shape, int [][][]trans, int i) {
        trans[i] = rotate(shape,0)     ;i++;
        trans[i] = rotate(trans[i-1],0);i++;
        trans[i] = rotate(trans[i-1],0);i++;
        trans[i] = rotate(trans[i-1],0);i++;
        return i;
    }

    // Rotate shape about y axis
    int [][] rotate (int [][] shape, int axis) {
        int[][] rot = new int[shape.length][];
        for (int i=0;i<shape.length;i++) {
            rot[i] = rotate(shape[i], axis);
        }
        return rot;
    }

    // Block rotators
    int [] rotate (int [] block, int axis) {
        switch (axis) {
            case 0: return new int[] { block[0], block[2],-block[1]};
            case 1: return new int[] {-block[2], block[1], block[0]};
            case 2: return new int[] { block[1],-block[0], block[2]};
        }
        throw new IllegalArgumentException();
    }

    // Return all positions the shape could occupy in the cube
    void addPositions (Set<Long> set, int [][] shape) {

        int xMin = min(shape,0);
        int xWidth = max(shape,0) - xMin + 1;
        int yMin = min(shape,1);
        int yWidth = max(shape,1) - yMin + 1;
        int zMin = min(shape,2);
        int zWidth = max(shape,2) - zMin + 1;

        // Add all encoded positions into array
        for (int x=0;x<=SIDE-xWidth;x++) {
            for (int y=0;y<=SIDE-yWidth;y++) {
                for (int z=0;z<=SIDE-zWidth;z++) {
                     set.add(encode(shift(shape,x-xMin,y-yMin,z-zMin)));
                }
            }
        }
    }

    // Return shape shifted by x,y,z amounts
    int [][] shift(int [][] shape, int x, int y, int z) {
        int[][] shift = new int [shape.length][];
        int i=0;
        for (int [] block : shape) {
            shift[i++] = new int[] { block[0]+x, block[1]+y, block[2]+z };
        }
        assert i==shape.length;
        return shift;
    }

    // Render blocks as SVG
    public static void toSVG(int[][][] soln) throws Exception {

        Writer out =  new FileWriter("bedlam.svg");

        // Write out header section
        out.write("<svg width='1000' height='1000' viewBox = '0 0 5000 5000' xmlns='http://www.w3.org/2000/svg' xmlns:xlink='http://www.w3.org/1999/xlink'>\n");
        out.write("<defs>\n");
        out.write("<polygon id='ppanel' points='0 0, 0 100, 86.6 50, 86.6 -50'/>\n");
        out.write("<g id='box'>\n");
        out.write("<use xlink:href='#ppanel' transform='rotate(-120)'/>\n");
        out.write("<use xlink:href='#ppanel' transform='rotate(0)'/>\n");
        out.write("<use xlink:href='#ppanel' transform='rotate(120)'/>\n");
        out.write("</g>\n");
        out.write("<g id='block'>\n");
        out.write("<use xlink:href='#box' style='fill:pink;'/>\n");
        out.write("</g>\n");
        out.write("<g id='topcube'>\n");
        out.write("<use xlink:href='#box' style='stroke-dasharray:1,3;' transform='translate(0,100),scale(4),translate(0,-100)' />\n");
        out.write("</g>\n");
        out.write("<g id='cube'>\n");
        out.write("<use xlink:href='#box' style='stroke-dasharray:1,3;' transform='rotate(180),translate(0,-100),scale(4),translate(0,100)' />\n");
        out.write("</g>\n");

        // Write shape definitions
        for (int b=0;b<soln.length;b++) {
            int [][] shape = soln[b];
            out.write("<g id='block" + b + "'>\n");
            for (int x=SIDE-1;x>=0;x--) {
                for (int y=SIDE-1;y>=0;y--) {
                    for (int z=0;z<SIDE;z++) {
                        for (int [] block : shape) {
                            if ((block[0]==x) && (block[1]==y) && (block[2]==z)) {
                                float tx = 86.60254f * (x-y);
                                float ty = -50.0f    * (x+y) - 100f * z;
                                out.write("<use xlink:href='#block' transform='translate(" + tx + "," + ty + ")'/>\n");
                            }
                        }
                    }
                }
            }
            out.write("</g>\n");
        }

        // Close of definitions and open layout graphic
        out.write("</defs>");
        out.write("<g style='fill:none;stroke:black;'>");

        // Layout blocks
        float x = 0f;
        float y = -200f;
        for (int b=0;b<soln.length;b++) {

            if ((b % 5) == 0) {
                y+=1000;
                x=800f;
            } else {
                x+=800f;
            }

            out.write("<use xlink:href='#cube' x='" + x + "' y='" + y + "'/>\n");
            out.write("<use xlink:href='#block" + b + "' x='" + x + "' y='" + y + "'/>\n");
            out.write("<text x='" + x + "' y='" + (y+200) + "' font-size='50'>" + b + "</text>\n");

        }

        // Close layout graphic and file
        out.write("</g>");
        out.write("</svg>");
        out.close();
        System.out.println("Sample solution output to bedlam.svg");
    }

    public static void main(String[] args) throws Exception {

        // Output SVG representation
        toSVG(SOLN);

        long t = System.currentTimeMillis();
        Bedlam b = new Bedlam();
        System.out.println("Bedlam created in " + (System.currentTimeMillis()-t) + "ms");
        if (true) b.search2();
        System.out.println("Complete " + (System.currentTimeMillis()-t) + "ms");
    }

}
